### ISSUE TITLE
Tray icon looks blurred/pixelated on KDE 

### ISSUE THREAD

**Issue by z3ntu** [telegramdesktop/tdesktop#2230]
As you can see in the screenshot, the Telegram Desktop icon looks very blurred in the system tray.
I am using KDE 5.6.5 and the newest version of Telegram Desktop.
![tdesktop](https://cloud.githubusercontent.com/assets/3768500/16614078/8cd3a1f4-436f-11e6-9a76-5c87164b3104.png)

It would also be good, if the Telegram Desktop tray icon would be monochrome for KDE but this is not important at all.

**Comment by xuzhao9** [comment]
Hello,
I am the packager of telegram-desktop on openSUSE and on my machine this problem does not happen.
Here is my [spec file](https://build.opensuse.org/package/view_file/home:nuklly/telegram-desktop/telegram-desktop.spec?expand=1).
Maybe it is useful to install Telegram icon into hicolor icon folder?
- Xu

**Comment by z3ntu** [comment]
I am using Arch Linux, not openSUSE (if you meant that). I downloaded the Binary off the official website and ran it.

**Comment by xuzhao9** [comment]
I believe for Arch Users, AUR has a package that integrates into the system better.
See: https://aur.archlinux.org/packages/telegram-desktop/

**Comment by z3ntu** [comment]
I have now installed the AUR package (compiled myself) and it looks even worse imho
![spectacle t16379](https://cloud.githubusercontent.com/assets/3768500/16646712/e5ba128c-442b-11e6-9a11-7142bca2b45b.png)

**Comment by xuzhao9** [comment]
Here is my screenshot:
![image](https://cloud.githubusercontent.com/assets/502017/16655287/9dfe64ce-4427-11e6-8243-1a4d278f1b48.png)

I don't think it is blurred on my system though.

**Comment by z3ntu** [comment]
The taskbar is like that for me (and when I make the taskbar bigger it is still in good resolution):
![spectacle tj8842](https://cloud.githubusercontent.com/assets/3768500/16656055/6a4b555c-445d-11e6-992c-aa156d2f7a62.png)

**Comment by gustawho** [comment]
The thing is, Telegram use PNG files for its images, while most of the tray icons are SVG (or compressed SVG's, *.svgz, in the case of Plasma, I think).

**Comment by alleehol** [comment]
I see the same problem when I resize the panel higher than the default:

Kubuntu 16.04 default (Plasma 5.5.5, KF5 5.18) Desktop, Intel Gfx, 27" 2560x1440 (std res)
Kubuntu 16.04 +ppa (Plasma 5.6.5, KF5 5.23) Laptop, Intel Gfx, 15" 1366x 768 (low res)

On the other hand I use the 'Icon-only' Taskmanger and there even a biiig Telegram Icon is crisp and sharp!  The taskmanager use the icon from the desktop file and it's a 256x256 png.

$ grep Icon= ~/Telegram/Telegram.desktop 
Icon=~/.TelegramDesktop/tdata/icon.png
$ file ~/.TelegramDesktop/tdata/icon.png 
/home/allee/.TelegramDesktop/tdata/icon.png: PNG image data, 256 x 256, 8-bit/color RGBA, non-interlaced

As I use plasmas default breeze theme, I've also
/usr/share/icons/breeze/apps/48/telegram.svg
on my system.

When I start Telegram in a konsole I get:
$ Telegram/Telegram
Fontconfig error: "/etc/fonts/conf.d/10-scale-bitmap-fonts.conf", line 72: non-double matrix element
Fontconfig error: "/etc/fonts/conf.d/10-scale-bitmap-fonts.conf", line 72: non-double matrix element
Fontconfig warning: "/etc/fonts/conf.d/10-scale-bitmap-fonts.conf", line 80: saw unknown, expected number
Fontconfig error: "/etc/fonts/conf.d/10-scale-bitmap-fonts.conf", line 72: non-double matrix element
Fontconfig error: "/etc/fonts/conf.d/10-scale-bitmap-fonts.conf", line 72: non-double matrix element
Fontconfig warning: "/etc/fonts/conf.d/10-scale-bitmap-fonts.conf", line 80: saw unknown, expected number

maybe that's telegram 'fault', maybe it's a KDE systray panel bug for resize events? Has anymore Plasma 5.7 running?  There the systray is rewritten.

**Comment by z3ntu** [comment]
@alleehol Yes, I have KDE 5.7.1 installed and it looks like on the attached screenshot


![image](https://cloud.githubusercontent.com/assets/3768500/16888065/12d8ba44-4ade-11e6-9f4c-e5d40030df7b.png)

**Comment by alleehol** [comment]
@z3ntu look good!  What happens when you make the panel higher (smaller).  Telegram systray icon sharp and crisp for all vertical sizes?

**Comment by z3ntu** [comment]
The tray icon stays blurry as it is, but the taskbar icon looks very good (probably high-res png or svg).

**Comment by Tynach** [comment]
This didn't used to be the case. It used to be crisp; I remember the text indicating how many messages I missed also being crisp. But it's not crisp anymore :/

Tried disabling XEmbedSNIProxy, and there was no change.

**Comment by z3ntu** [comment]
I just found the line

> Jul 29 12:18:25 hp-praktikum-luca xembedsniproxy[671]: kde.xembedsniproxy: Scaling pixmap of window 52428807 "telegram-desktop" from w*h 22 22

in my log. That's the problem probably.

**Comment by alleehol** [comment]
hmm, AFAIU xembed deprecated  https://blog.martin-graesslin.com/blog/tag/systray/  There was a time KDE's plasma did not support it.  If telegram supports both methods xembed and status notifier item,
status notifier should be tried first, that should solve the problem (and would work with wayland too).

Achim

**Comment by Tynach** [comment]
@alleehol, @z3ntu, the problem persists when I disable Xembed. I also do not have lines like that in my Telegram log, so this is not the root of the problem.

I actually suspect that those of you having a blurry icon are perhaps using Xembed, while those of us who have a horribly pixelated icon are using libnotify.

But I don't know. All I know is that I'm not using Xembed, and the problem persists.

**Comment by Tynach** [comment]
It seems that [Telegram/SourceFiles/platform/linux/main_window_linux.cpp][source] is responsible for handling the system tray icon. It appears to create new icons for each time it's changed, which makes sense for supporting the message count feature.

In this file you can see that they support 3 different styles of system tray icons:

1. `QSystemTrayIcon`
2. `AppIndicator`
3. `GtkStatusIcon`

Of those three, the only one that actually changes the value of the variable `_trayIconSize` is `GtkStatusIcon`. Neither `AppIndicator` nor `QSystemTrayIcon` change that variable, yet all of them use it for generating the icon size.

The starting value for that variable is 22, and indeed if I go to my `~/.TelegramDesktop/tdata/ticons/` folder I can see that the icons it generated are 22x22 pixels large.

Paging the author of this code, @john-preston. Perhaps he can give some insights and ideas.

[source]: https://github.com/telegramdesktop/tdesktop/blob/master/Telegram/SourceFiles/platform/linux/main_window_linux.cpp

**Comment by john-preston** [comment]
@Tynach Which size should it be? I was taking a size to look fine on the systems where I've tested the build.

**Comment by Tynach** [comment]
@john-preston, in KDE the panels can be dynamically resized at runtime, so ideally the code would use whatever size is requested. I don't know exactly what that size might be (and it will change depending on the user's preferences), but with the default panel (40px tall with fonts set to 96 DPI) the icons appear to be 32x32 pixels.

You do have a resize handler for `GtkStatusIcon`. Is it possible for you to write such a handler for `AppIndicator` or `QSystemTrayIcon`? Failing that, at least a function or method that detects what the requested icon size is?

It also seems that [QSystemTrayIcon will use AppIndicator or XEmbed][one stone] on Linux, choosing whichever one is available/works best. So perhaps you could just use that and not have separate code for AppIndicator (especially since Telegram has Qt and whatnot compiled statically).

It also apparently supports Windows and OS X, so you could potentially remove this from platform specific implementation and simply have it as the default code for all platforms. However, I don't know how well integrated it looks on other OSes.

I do see in the documentation that 22px is considered the standard size for Linux (X11) desktops. I suppose in this way that KDE is the oddball; though I swear this worked fine in a previous build. Weird.

At any rate, apparently [`QIcon`] has a lot of documentation too, including a way to choose which 'Icon Engine' you use; which might be a simpler way of choosing between XEmbed and AppIndicator.

I also see the `QIcon::actualSize()` method that tells you the dimensions of a pixmap given a certain size, which it claims will never be larger than the requested size but can be smaller. Perhaps you could put in a ridiculous size like '512', and then use the 'actual size' it returns as the size of the icon you generate with `_trayIconImageGen()`.

I'd look around for more details (so far I've only dabbled with making little useless 'toy' programs using Qt, and kinda want to try implementing my own system tray icon in a program just to see how it all works), but I'm currently packing for a trip.

If I'd already dabbled in this corner of Qt, and wasn't about to leave for a week and a half, I'd offer to try writing the code myself and submit a pull request.

[one stone]: http://doc.qt.io/qt-5/qsystemtrayicon.html#details
[`QIcon`]: http://doc.qt.io/qt-5/qicon.html

**Comment by john-preston** [comment]
@Tynach I have a resize handler for gtk status icon because it offers such a signal. I don't see anything like that for appindicator: http://people.canonical.com/~dpm/api/devel/GIR/C/AppIndicator-0.1.html

**Comment by Tynach** [comment]
@john-preston, it seems that `QIcon` has a method where you can add multiple of the same icon, but each a different size, and the system will use whichever one fits the best. This might be a better approach.

I talk about that and a few other things lower down in my post up there.

**Comment by z3ntu** [comment]
I also found

> Jul 31 12:12:21 g550jk kdeinit5[1308]: Registering "org.kde.StatusNotifierItem-14215-1/StatusNotifierItem" to system tray
Jul 31 12:12:21 g550jk plasmashell[1352]: Registering "org.kde.StatusNotifierItem-14215-1/StatusNotifierItem"

in my logs. This is in the normal systemd journal (journalctl)

**Comment by davidedmundson** [comment]
Plasma dev here.

>hmm, AFAIU xembed deprecated https://blog.martin-graesslin.com/blog/tag/systray/ There was a time KDE's plasma did not support it.

That's not quite true, it is supported now, however you won't be using it as you have AppIndicator support.

>It also seems that QSystemTrayIcon will use AppIndicator or XEmbed

That's correct from Qt5.7, and probably worth doing, but it won't help fix anything, as you can see below.

>AppIndicator

There's no way to get the size the host is displaying at. It's not part of the spec, so whatever abstraction layer you use it simply won't work. (for one thing there may be N hosts).
Even if you have a QIcon with multiple sizes, it can still only send one image as an IconPixmap field.

Rather than passing the icon as a pixmap over DBus, it's more typical to use the IconName + IconPath. It's faster, simpler, and it avoids the resolution problems as it's the client loading the icon. 
The iconPath supports the full XDG icon directory convention with the different folders per sizes, or SVG.

However, from what I can see you're deliberately doing it as a pixmap as you're manipulating the icon to display a badge with a number on it?

At which point your only option available is to just send a huge icon to begin with and let the indicator host scale it down. 

Long term you might want to start a discussion about exanding the specification so that host side overlays are a thing. I think Unity has a specification for that(?)
Or we expand the SNI spec to support arrays in IconPixmaps

**Comment by john-preston** [comment]
@davidedmundson Yes, I change the icon placing different badges. I thought about sending a large enough pixmap, though it will generate slowly (like 512x512), then scale back to small and will have worse quality (which is critical for small icons with very small counter of new messages — they're best displayed when prepared using the needed size from the beginning). Of course it doesn't work well, if they're then upscaled. Perhaps sending a large enough pixmap should be considered at some of the future versions.

**Comment by alleehol** [comment]
If supported: Maybe SVG is then the simpler solution? Substituting the Number in the SVG should be quick, and the scaling is done only once for earch systray instance  (FWI I tried a inch high panel and at least the KDE systray impl. has a max size for the icons of ~ 25 px.)   
![screenshot-high-plasma-panel](https://cloud.githubusercontent.com/assets/5500644/17276138/fc2ed316-571f-11e6-8a90-b57dd6e4fd11.png)

If possible, this should also solve 2 or more Panel with systrays (1st small, 2nd much taller) because the SVG 'supports' several sizes.

**Comment by davidedmundson** [comment]
@alleehol 
Telegram doesn't control how Plasma renders it.
and Plasma doesn't know it's meant to substitute things.

They're running in different processes, and the only communication between them expects a single pixmap or a text based icon name.

**Comment by sl1pkn07** [comment]
why no use StatusNotifier instead of accient Xembed?

http://www.notmart.org/misc/statusnotifieritem/

**Comment by AndydeCleyre** [comment]
While I lose the mute signifier and the message count, and the icon is colored instead of mono as is more fitting for a plasma icon, I can get a non-pixelated icon up there as follows:

```bash
cd ~/.TelegramDesktop/tdata/ticons
rm -r *
cp /usr/share/pixmaps/telegram.png ./ico_22.png
cp ico_22.png icomute_22_0.png
cp icomute_22_0.png icomute_22_1.png
for f in {1..2000}; do ln -s ico_22.png ico_22_$f.png; done
for f in {2..2000}; do ln -s icomute_22_1.png icomute_22_$f.png; done
```

That's a 256x256 image.

**Comment by AndydeCleyre** [comment]
Alright, I'm finding this script works well for me. Just missing the message count.

```bash
#!/bin/bash

cd ~/.TelegramDesktop/tdata/ticons
rm -r *
cp /usr/share/pixmaps/telegram.png ./ico_22.png

# blue -> orangered:
convert ico_22.png -modulate 100,100,0 ico_22_1.png

# darken
convert ico_22.png -modulate 50,100,100 icomute_22_0.png

# darken orangered
convert ico_22_1.png -modulate 50,100,100 icomute_22_1.png

for f in {2..2000}; do ln -s ico_22_1.png ico_22_$f.png; done
for f in {2..2000}; do ln -s icomute_22_1.png icomute_22_$f.png; done
```

**Comment by AndydeCleyre** [comment]
I just found that the source icon used for the above script is not included in the Ubuntu Telegram package available from http://ppa.launchpad.net/atareao/telegram/ubuntu . The source icon is available at https://raw.githubusercontent.com/telegramdesktop/tdesktop/master/Telegram/Resources/art/icon256.png .

**Comment by sl1pkn07** [comment]
better

https://upload.wikimedia.org/wikipedia/commons/d/dd/Telegram_alternative_logo.svg

**Comment by sl1pkn07** [comment]
why this issue is closed? still get a pixelated systray icon

**Comment by hesediel84** [comment]
what about other icons with the number of messages? i changed the ico22 and solved the problem of the main icon. But the icon with the number of messages in red still pixelated

**Comment by sl1pkn07** [comment]
i think is possible change this Behavior from this code

https://github.com/telegramdesktop/tdesktop/blob/0d7af044e91452f7bd0db39c6eb9316e16d6071e/Telegram/SourceFiles/platform/linux/main_window_linux.cpp#L87

**Comment by Getron** [comment]
Even here, on 0.10.10 on Linux Manjaro 16.08 KDE 5.7 Rolling Arch Release, the icon is not fixed, and looks blury, ugly.

Please, fix this, maybe a bigger, or more detailed, more pixel icon would be fix the problem on linux KDE. All other messengers have fixed the icons, just Telegram seems to be blured, pixelated.

**Comment by mrzzzrm** [comment]
Similar problem here, additionally the icon background has the wrong color

![screenshot_20161009_203156](https://cloud.githubusercontent.com/assets/2679018/19222742/aba03936-8e5f-11e6-9c14-e7403f2bc514.png)

**Comment by z3ntu** [comment]
@mrzzzrm For me it's pixelated the same way but it doesn't have that weird background color...

**Comment by DaavisGit** [comment]
Icon problems solved when i installed nvidia driver 367.57 on kubuntu

**Comment by z3ntu** [comment]
I still noticed the issue with 375.10

**Comment by sl1pkn07** [comment]
nvidia (or video) drivers make nothing in this issue

**Comment by DaavisGit** [comment]
@sl1pkn07 well.. my tray icon fixed after the driver install. Maybe there is no direct connection to it, but there must be some kind of connection

**Comment by sl1pkn07** [comment]
nope. if the image of the systray is small, if upscale, the image gets pixelated

set bigger image to the systray solve the issue

video drivers only apply filters to the image (antialiasing) displayed

**Comment by DaavisGit** [comment]
@sl1pkn07 all I did was install display drivers.. How would you explain icon fixing on itself?

**Comment by sl1pkn07** [comment]
see my last statement

**Comment by fliker09** [comment]
Same here, Kubuntu 16.04.1, badly pixelated :\ I don't want to lose notification badge. Is there a solution?

**Comment by z3ntu** [comment]
Did anyone else notice an improvement in 1.0.13?

**Comment by AndydeCleyre** [comment]
This is how I'm getting around it these days:

```bash
panel_icons="/usr/share/icons/Papirus-Dark/24x24/panel"
icon_size="512"
main_color="#c2b790"
accent_color="#a94d37"
main_color_old="#d3dae3"
accent_color_old="#5294e2"

# hardcode-tray -a -cc "${main_color_old} ${main_color}" -cc "${accent_color_old} ${accent_color}" --theme Papirus-Dark
# ^ This currently still makes the icons too small. So instead:

tg_tmp="/tmp/tg_svgs"
mkdir -p ${tg_tmp} ${tg_icons}
for i in '_22':'' '_22_1':'attention-' 'mute_22_0':'' 'mute_22_1':'mute-'; do
    ico=${i%:*}; name=${i#*:}
    sed "s/${main_color_old}/${main_color}/gi" ${panel_icons}/telegram-${name}panel.svg > ${tg_tmp}/telegram-${name}panel.svg
    sed -i "s/${accent_color_old}/${accent_color}/gi" ${tg_tmp}/telegram-${name}panel.svg
    inkscape -z -e ${tg_icons}/ico${ico}.png -w ${icon_size} -h ${icon_size} ${tg_tmp}/telegram-${name}panel.svg &>/dev/null
done
for f in {2..2000}; do
    ln -sf ${tg_icons}/ico_22_1.png ${tg_icons}/ico_22_${f}.png
    ln -sf ${tg_icons}/icomute_22_1.png ${tg_icons}/icomute_22_${f}.png
done
```

**Comment by sl1pkn07** [comment]
```bash
#!/usr/bin/bash
source_image="test.svg"
counter_image="red_point.png"
final_size=256
counter_size=180
font=Ubuntu # see 'identify -list font' for obtain the name of the font
font_size=78
messages=10

# Create the image ico_22_0.png and icomute_22_0.png symlink
convert -density 200 -background transparent -resize ${final_size} ${source_image} ico_22_0.png
ln -sf ico_22_0.png icomute_22_0.png

# Create the image for message counters
for i in $(seq 1 $messages); do
  counter="\"_counter_\""
  # draw the number of the message into the counter image
  convert -resize ${counter_size} ${counter_image} -gravity Center -font ${font} -pointsize ${font_size} -fill white -draw "text 0,0 ${counter//_counter_/$i}" counter.png
  # merge the counter image into the background image
  composite -gravity SouthEast counter.png ico_22_0.png ico_22_${i}.png
  # symlink ico*foo to icomute*foo
  ln -sf ico_22_${i}.png icomute_22_${i}.png
  # remove the counter image
  rm counter.png
done
```

test.svg: https://cgit.kde.org/breeze-icons.git/tree/icons-dark/apps/48/telegram.svg
red_point.png: ![red_point](https://user-images.githubusercontent.com/462213/29049558-f97e6dca-7bd6-11e7-8207-76fe28d9ccd9.png)

result:

10 messages: ![ico_22_10](https://user-images.githubusercontent.com/462213/29049955-3b436cb8-7bd9-11e7-9214-9114fad39d56.png)
999 messages  ![ico_22_999](https://user-images.githubusercontent.com/462213/29049963-4524a6ac-7bd9-11e7-9150-06ee6ab45c03.png)
1000 messages ![ico_22_1000](https://user-images.githubusercontent.com/462213/29049968-4d4ea6ac-7bd9-11e7-8f19-391179406c08.png)
9999 messages  ![ico_22_9999](https://user-images.githubusercontent.com/462213/29049971-55dc0c4c-7bd9-11e7-91fe-33e91fe6bdc6.png)


more or less, like the original

anyone can test it?

i think 256x256 is a good resolution for the trayicon

**Comment by ElvisPresly** [comment]
I will just add the picture here)
![2017-08-23 19-11-40](https://user-images.githubusercontent.com/18265958/29650342-2a12c17e-88a3-11e7-8873-997d1ac4890e.png)
![2017-08-23 19-12-43](https://user-images.githubusercontent.com/18265958/29650343-2a2fcd28-88a3-11e7-8875-f4ae49f1c1b0.png)

**Comment by SovRedBit** [comment]
I have the same problem like @mrzzzrm has, telegram background color is pink on Kubuntu 16.04

**Comment by alexey-khabulyak** [comment]
have the same issue. kde neon. Kde 5.14.2
with autostart
![screenshot_1](https://user-images.githubusercontent.com/19747881/48072996-cba73880-e1ee-11e8-9196-aad95aa75b00.png)
with manual startup
![screenshot_2](https://user-images.githubusercontent.com/19747881/48073006-cf3abf80-e1ee-11e8-867f-a8783ff1fe73.png)

**Comment by ant-** [comment]
I've faced with the same problem on Ubuntu 18.04 with Plasma version 5.12.7. 
I have 2 machines running Ubuntu with KDE. The only difference the one is a pure Kubuntu (#1), and the other one is a just Ubuntu with kubuntu-desktop (#2) installed separately.
Screenshots:
Machine #2
[![enter image description here][1]][1]

[![enter image description here][5]][5]
[![enter image description here][6]][6]

Machine #1

[![enter image description here][2]][2]

[![enter image description here][3]][3]
[![enter image description here][4]][4]

As you can see in both cases telegram icon is blured. Skype is just for reference, but definitely there is something wrong with it too as well. So, maybe it is not a telegram or skype issue.

  [1]: https://i.stack.imgur.com/RmNDY.png
  [2]: https://i.stack.imgur.com/5HsyN.png
  [3]: https://i.stack.imgur.com/k2gSz.png
  [4]: https://i.stack.imgur.com/0nXdk.png
  [5]: https://i.ibb.co/Njpg8tR/Screenshot-20181222-111145.png
  [6]: https://i.ibb.co/k07H48P/Screenshot-20181222-111221.png

**Comment by yuritoledo** [comment]
any update?

**Comment by AndydeCleyre** [comment]
FWIW I'm now using the following script to replace these icons with some from the Papirus icon theme. This method does not reflect the number of messages.

```bash
#!/bin/bash
set -e

plain_color=`python3 -c "print(''.join(hex(int(i))[2:] for i in '$(kreadconfig5 --group Colors:Window --key ForegroundNormal)'.split(',')))"`
alert_color=`python3 -c "print(''.join(hex(int(i))[2:] for i in '$(kreadconfig5 --group Colors:Window --key DecorationFocus)'.split(',')))"`

cd ~/.local/share/TelegramDesktop/tdata/ticons
rm -rf *.png

cp /usr/share/icons/Papirus/24x24/panel/telegram-* ./
sed -i "s/dfdfdf/$plain_color/" *.svg
sed -i "s/4285f4/$alert_color/" *.svg
sed -i "s/$plain_color/$alert_color/" telegram-attention-panel.svg

inkscape -z -e icon_22_0.png -w 256 -h 256 telegram-panel.svg
inkscape -z -e icon_22_1.png -w 256 -h 256 telegram-attention-panel.svg
inkscape -z -e iconmute_22_1.png -w 256 -h 256 telegram-mute-panel.svg
rm -rf *.svg

ln -s icon_22_0.png iconmute_22_0.png
for f in {2..2000}; do ln -s icon_22_1.png icon_22_$f.png; done
for f in {2..2000}; do ln -s iconmute_22_1.png iconmute_22_$f.png; done
```

![image](https://user-images.githubusercontent.com/1787385/59712632-9aa6d180-91db-11e9-8290-a3db1e3edfbc.png)

![image](https://user-images.githubusercontent.com/1787385/59712717-c0cc7180-91db-11e9-8518-87d594442cea.png)

**Comment by VictorQueiroz** [comment]
Seriously? All that and nobody can fix a "low resolution icon on tray" problem? That's just sad. I hope to find a solution on an official TG desktop release someday.

**Comment by Aokromes** [comment]
https://github.com/telegramdesktop/tdesktop/commit/3b4dfa138176b4d1e3cb14be1be638ebe2f52713
